Allow webpack to generate hard-coded stylesheets in development

nickautomatic opened the pull request:
This commit introduces a `USE_STYLESHEETS` environment variable that allows us to (optionally) generate hard-coded stylesheets in development.

We will separately update webpack's configuration in the main app to allow us to make use of this change.

The motivation here is that we currently have no easy way to test the styling of components that import their own styles when JavaScript is disabled. This will become more of a problem the more components become responsible for importing their own styles, and we will also need it in order to test the "old browser" (ie. largely no-JS) experience once we've started to [cut the mustard](https://github.com/Futurelearn/futurelearn/pull/2655).

We've tested this change in the main app on [a branch](https://github.com/Futurelearn/futurelearn/commits/nf-sm-import-styles) that points at this version of webpack-config, and it doesn't seem to break anything! Is there any reason not to go ahead and merge this?

thomasmarshall commented:
Hey @nickautomatic – do you know why we use the `style-loader` at all in development? Is it that it's faster or that it has some kind of live reload functionality?

nickautomatic commented:
@thomasmarshall and @wannabefro - I tried a couple of times to get feedback on this in #tech with no luck, so requesting a code review from you because I think you both know enough about webpack to check this... The main thing I'd like is just a double check that this is safe to merge!

thomasmarshall commented:
@nickautomatic I recommend rebasing this and running it again locally. I recently tried to replace `style-loader` entirely with `MiniCssExtractPlugin` and it increased memory usage significantly, such that the build process crashed fairly consistently.

nickautomatic commented:
@thomasmarshall Oh, good point - this is pretty stale now... Sorry, probably should have rebased before posting the request for feedback! 🙂 I'll do that. I might give you a shout if it's not clear how to resolve the conflicts.

nickautomatic committed:
Allow webpack to generate hard-coded stylesheets in development

Trello: https://trello.com/c/YTi3Gjfz

In development, we inject CSS Modules stylesheets via JavaScript.
This is to keep the build time low, because the CSS extract plugin
is quite slow.

The downside of this is that these stylesheets don't get loaded if JS
is disabled.

However, as we move towards using CSS Modules more widely, and
especially once we start cutting the mustard (ie. not serving JS to
older browsers), it's important that we should be able to test the site
with JS disabled and have it look as it would in production.

This commit updates the webpack config to allow use of a
`USE_STYLESHEETS` environment variable that allows us to tell webpack to
run as if in production (ie. extract JS-imported stylesheets and include
them in a way that is not dependent on JS).

Initially we can use it directly from the command line:

    USE_STYLESHEETS=true ./bin/webpack

...but there might be ways we could make this easier to use.

Co-authored by: wannabefro <sam.mctaggart@gmail.com>

nickautomatic commented:
I've now rebased this and tested again locally and it's still working (ie. you can run `USE_STYLESHEETS=true ./bin/webpack` and everything should look ok with JS turned off).

I'm wondering if the commit message is quite telling the full story now, because the bit where I'm saying:

> This is to keep the build time low, because the CSS extract plugin is quite slow.

...no longer feels like a complete explanation, since as of 550b55a484efb2cb4e1c2c8b019faed8dd1bf6de we _are_ now using the CSS extract plugin for all but our CSS Modules styles. @thomasmarshall, it sounds like this is a thing that just became necessary for practical reasons - is that right?

thomasmarshall commented:
> the rationale here for why we're still avoiding it with CSS Modules (ie. it's slower than the alternative, and we don't absolutely need it here) is still basically correct... is that right?

I think so. The other reason was that if we wanted hot module replacement we'd have to use `style-loader`. However, `MiniCssExtractPlugin` _does_ support some kind of HMR: https://github.com/webpack-contrib/mini-css-extract-plugin#hot-module-reloading-hmr

nickautomatic commented:
@thomasmarshall Ok, great, in that case I think the commit message can stay as it is.

Is it ok if I go ahead and merge this now, do you think?
